Assets/scripts/Inputs/InputManager.cs

```diff
@@ -1,79 +1,175 @@
 ﻿using System.Collections;
 using System.Collections.Generic;
+using Interfaces.ServiecesInterface;
+using Player;
+using Replay_Service;
+using StateMachines;
 using UnityEngine;
 using UnityEngine.SceneManagement;
 
-public class InputManager : Singleton<InputManager>
+public class InputManager : IInputManager
 {
-    public Dictionary<Controls, InputData> playerInput = new Dictionary<Controls, InputData>();
+    //public int initialFame, pauseFrame, resumeFrame;
+    public Dictionary<Controls, Queue<InputData>> playerInput = new Dictionary<Controls, Queue<InputData>>();
+    public GameObject InputControllersPrefab, referenceInput;
     public Dictionary<Controls, List<InputComponent>> inputComponents = new Dictionary<Controls, List<InputComponent>>();
-
-
+    bool userControls = true, replay = false, pause = false;
     // Start is called before the first frame update
     // Update is called once per frame
-    private void OnLevelWasLoaded(int level)
+    public InputManager()
+    {
+        InputControllersPrefab = Resources.Load<GameObject>("Inputs");
+        SceneManager.sceneLoaded += OnLevelLoaded;
+        ServiceLocator.Instance.get<IStateManager>().OnStateChanged += GameStateChanged;
+        GameApplication.Instance.OnPlayerSpawn += AddPlayerListener;
+    }
+    public void AddPlayerInputData(Controls controls)
+    {
+        playerInput.Add(controls, new Queue<InputData>());
+
+    }
+    public void SetQueue(Dictionary<Controls, Queue<InputData>> Pinput)
     {
-        if (SceneManager.GetActiveScene().name == "GameScene")
+        playerInput = Pinput;
+    }
+    public void EnqueueData(InputData inputdata, Controls controls)
+    {
+        if (playerInput.ContainsKey(controls))
         {
+            //ServiceReplay.Instance.RecordInput(inputdata, controls);
+            playerInput[controls].Enqueue(inputdata);
+        }
+    }
+    public void Update()
+    {
+        if (!pause)
+        {//Debug.Log(replay);
+            foreach (Controls controls in inputComponents.Keys)
+            {
 
-            GameApplication.Instance.OnPlayerSpawn += AddPlayerListener;
+                foreach (InputComponent inputComponent in inputComponents[controls])
+                {
+                    if (playerInput[controls].Count > 0)
+                    {
+                        //
+                        // if (replay)
+                        // {
+                        //         Debug.Log("current: " + (Time.frameCount - initialFame) + "input: " + Instance.playerInput[controls].Peek().frame);
+                        // }
+                        if ((ServiceLocator.Instance.get<IFrameService>().GetFrame()) == (playerInput[controls].Peek().frame))
```

Comment:
cache the frame service since you need it in update and loops

Reply:
done